dashboard/design/survey_summary_design.md

```diff
@@ -0,0 +1,161 @@
+We don't have to start from step 1 (Ingest), we can start from step 3 (Summarize) and create adapters to connect it to the current pipeline.
+
+Specifically for CSF survey work, we will create an input adapter to retrieve survey data from the current data model, and an output adapter to convert summary result to format that the current UI view understands. Looking at it from a different angle, these adapters are just concrete implementations of Retriever and Decorator.
+
+
+# Recommendation
+Option 3. Build a new generic survey pipeline, starting from step 3 (Summarize module) first and create adapters to plug it in the current pipeline. This approach allows us to support 2019 CSF survey sooner by taking advantage of the existing work. Then, gradually build out the rest of the pipeline later.
+
+## Rollout plan
+1. Build 5 components just enough to support 2019 surveys: Retriever (for current db), Transformer, Mapper, Reducer and Decorator (for current UI view). They are components in step 3 and 4 of the above diagram.
+2. Build Fetcher (step 1) to download survey submissions from JotForm to our database.
+3. Build Parser, Indexer and Filter (step 4) to allow quick submission search.
+4. Build another implementation of Retriever (in step 3) which will read from the new db created in step 2 & 3.
+5. Build Presenter (in step 4) that can display result from a single query (e.g. summary results of 1 JotForm survey of a specific workshop)
+6. Build Decorator (in step 4) to provide enough information for the Presenter built in step 5.
+7. Switch 2019 surveys to use the new pipeline. It should then be completely independent from the older pipeline.
+8. All new surveys will use the new pipeline form this point forward.
+9. Migrate surveys using older pipeline to newer one when we think it's worth to do so.
+
+## 5 components to build for 2019 surveys
+Estimated Cost: 4.5 ± 1 days.\
+This estimation is for features with basic unit and integration tests. Thorough testing will take longer.
+
+1. Retriever
+  - Retrieve survey data from existing tables (`SurveyQuestion, WorkshopDailySurvey`) and convert data to `submission_content_array` format.
+  - Estimate: 2d
+2. Transformer
+  - Convert survey data from submission centric to question centric. At high level: {submission_id, submission_data} --> {question_name, question_and_answer_data}.
+  - Estimate: 0.5d
+3. Mapper
+  - Go through question data, send it to suited Reducer based on question name and type. E.g. text -> text collector, number -> histogram/count/avg reducers.
+  - Estimate: 0.5d
+4. Reducer(s)
+  - Produce a summary result from list of input question data. E.g. text collector, histogram, avg, count, distribution reducers.
+  - Estimate: 0.5d
+5. Decorator
+  - Compile minimum information needed for the current UI view (`local_summer_workshop_daily_survey/results.jsx`) to display summary results.
+  - Estimate: 1d
+
+Note, we skip Modifier component because it's an optional feature.
```

Comment:
For some reason I had the idea that the Modifier was the whole point, because we currently have surveys we can't roll up together with the existing system (e.g. the Organizer Survey Results problem @clareconstantine is working through).  But you don't think we need it right away? Can you help me understand what you find valuable about this work without the Modifier?

Reply:
The main motivation of the new pipeline is to reduce the cost to summarize and present summaries for new surveys. It will support rolls up multiple surveys from multiple workshop also because it processes data at question_unique_name level, not at survey or workshop level.

I wrote a few more in the **Motivation** section. We could discuss this in person if it isn't clear yet.